macros/ui/niceTables.pl

```diff
@@ -3,835 +3,1201 @@ =head1 niceTables.pl
-						or ($headerrow[$i] == 1) and !(uc(${ $dataref->[$i][$j] }{header}) eq 'TD'))
-				{
-					${ $dataref->[$i][$j] }{tex} = '\bfseries ' . ${ $dataref->[$i][$j] }{tex};
-				}
+	# special user shortcut
+	$outHash{texalignment} = $userOptions{align}
+		if (defined($userOptions{align}) && !$userOptions{texalignment});
 
-				if (${ $dataref->[$i][$j] }{multicolumn} ne '') { $textable .= ${ $dataref->[$i][$j] }{multicolumn} }
-				$textable .=
-					${ $dataref->[$i][$j] }{texpre} . ' '
-					. ${ $dataref->[$i][$j] }{tex} . ' '
-					. ${ $dataref->[$i][$j] }{data} . ' '
-					. ${ $dataref->[$i][$j] }{texpost};
-				if (${ $dataref->[$i][$j] }{multicolumn} ne '') { $textable .= '}' }
-				$textable .= '&' unless ($j == $numcols[$i]);
-			}
-			$textable .= '\\\\';
-			if ($midrule[$i] == 1)                                                    { $textable .= '\midrule ' }
-			if ((($midrules == 1) or ($headerrow[$i] == 1)) and ($i != $#{$dataref})) { $textable .= '\midrule ' }
-		}
-		$textable .= '\bottomrule' . $endtabular;
-		$textable .= '\end{minipage}\par  \vspace{1pc}';
-	}    # and now if it is a Layout Table...
-	else {
-		my ($begintabular, $endtabular) = ('\begin{tabular}', '\end{tabular}');
-		if ($texalignment =~ /X/) {
-			($begintabular, $endtabular) = ('\begin{tabularx}{' . $Xratio . '\linewidth}', '\end{tabularx}');
+	# legacy misnomers
+	$outHash{horizontalrules} = $userOptions{midrules}
+		if (defined($userOptions{midrules}) && !$outHash{horizontalrules});
+	return \%outHash;
+}
+
+sub ParseAlignment {
+	my $alignment = shift;
+	$alignment =~ s/\R//g;
+
+	# first we parse things like *{20}{...} to expand them
+	my $pattern = qr/\*\{(\d+)\}\{(.*?)\}/;
+	while ($alignment =~ /$pattern/) {
+		my @captured    = ($alignment =~ /$pattern/);
+		my $replaceWith = $captured[1] x $captured[0];
+		$alignment =~ s/$pattern/$replaceWith/;
+	}
+
+	my @align = ();
+
+	# 0th entry is only for possible left border
+	# other entries have only right borders,
+	# the actual alignment is r, c, l, X, or p
+	# explicit width vertical rules from !{...}
+	# latex directives from >{...}
+	# we make an array of the tokens of type:
+	# r, c, l, X, |, !{...}, p{...}, >{...}
+	# this is complicated because of potential nested brackets
+	my @tokens             = ();
+	my $bracesregex        = qr/(\{(?>[^{}]|(?R))*\})/x;
+	my $bracecontentsregex = qr/((?>[^{}]|(??{$bracesregex}))*)/x;
+
+	# . at the end is to ensure we are whittling down $alignment at least a little
+	my $tokenspattern = qr/^([rclX\|]\s*|[!p>]\s*\{((??{$bracecontentsregex}))\}\s*|.)/;
+
+	$align[0] = { left => 0 };
+	$leftpattern = qr/^\s*(\|\s*|!\s*\{\s*\\vrule\s+width\s+([^}]*?)\})/x;
+	while ($alignment =~ $leftpattern) {
+		my $token = $1;
+		if ($token =~ /^\|/) {
+
+			# this counts how many | we have
+			$align[0]->{left} = 0
+				unless ($align[$i]->{left} && $align[$i]->{left} =~ /\d+/);
```

Comment:
I believe this should be
```suggestion
			$align[0]{left} = 0
				unless ($align[0]{left} && $align[0]{left} =~ /\d+/);
```
Note:  Once you access an array or hash by the first index or key, there is no need for further dereferencing via `->`.  It doesn't do anything but make the line longer.  An array or hash can not contain non-scalar values that don't need to be dereferenced, so it doesn't add clarity to the code either.